semantics/src/main/scala/aqua/semantics/expr/CallArrowSem.scala

```diff
@@ -25,16 +25,29 @@ class CallArrowSem[F[_]](val expr: CallArrowExpr[F]) extends AnyVal {
     at: ArrowType
   )(implicit
     N: NamesAlgebra[F, Alg],
+    T: TypesAlgebra[F, Alg],
     V: ValuesAlgebra[F, Alg]
-  ): Free[Alg, List[Call.Arg]] =
+  ): Free[Alg, (List[ValueModel], Option[Type])] =
     V.checkArguments(expr.funcName, at, args) >> variable
-      .fold(freeUnit[Alg])(exportVar =>
+      .fold(freeUnit[Alg].as(Option.empty[Type]))(exportVar =>
         at.res.fold(
           // TODO: error! we're trying to export variable, but function has no export type
-          freeUnit[Alg]
-        )(resType => N.define(exportVar, resType).void)
-      ) >> args.foldLeft(Free.pure[Alg, List[Call.Arg]](Nil)) { case (acc, v) =>
-      (acc, V.resolveType(v)).mapN((a, b) => a ++ b.map(Call.Arg(ValuesAlgebra.valueToModel(v), _)))
+          freeUnit[Alg].as(Option.empty[Type])
+        )(resType =>
+          N.read(exportVar, mustBeDefined = false).flatMap {
```

Comment:
I think it is a mistake that `read` is lifting an error if it is not defined. Maybe split it with two functions instead of adding a flag?

Reply:
Now it feels sound with all the other semantic rules. Let's live with it as it is for a while and then decide.